src/provider/provider_coarse.c

```diff
@@ -0,0 +1,1224 @@
+/*
+    Copyright (c) 2023 Intel Corporation
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+        http://www.apache.org/licenses/LICENSE-2.0
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "../src/memory_provider_internal.h"
```

Comment:
I think we should avoid using `*_internal.h` headers in provider/pool implementations. If there is something we need from those headers, it should probably be part of the public API.

IMO it should be possible to implement all our pools/providers without relying on internal details. If not, then it will be challenging to have any pool/provider implemented outside of this repo.